### PR TITLE
[nn] Allow eps=0 in batch_norm during eval mode

### LINKED ISSUE #173337: torch.nn.BatchNorm expects positive eps in eval mode
### 🐛 Describe the bug

torch 2.10.0 enforces `torch.nn.BatchNorm` to have positive eps https://github.com/pytorch/pytorch/blob/567eee90e8af4f51f9dba3034902ce4d0aa5ebd6/torch/nn/functional.py#L2843-L2844

In eval mode, `running_mean` and `running_var` are frozen and can be constant-folded and hence eps can be 0 

**Repro**
```python
import torch
import torch.nn as nn

batch_norm = nn.BatchNorm1d(
    num_features=512,
    eps=0,
    momentum=0.9
)

# eval mode
batch_norm.eval()

input_tensor = torch.randn(16, 512)  # (batch_size, num_features)
output = batch_norm(input_tensor)
```

There are couple of options here
1. Check `running_var` along with `eps` to avoid `divide-by-zero`
2. non-zero `eps` check only in case of training mode

option 1 is more general but comes with added complexity of checking `running_var`

### Versions

torch==2.10.0

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

### PR DESCRIPTION
**Pull request by bledden**
## Summary

`torch.nn.functional.batch_norm` currently requires `eps > 0` unconditionally, but in eval mode `running_mean` and `running_var` are frozen and can be constant-folded or pre-adjusted by the user. There are valid use cases where `eps=0` makes sense in eval mode (e.g., when running stats have been deliberately set to avoid division by zero).

This relaxes the check so that:
- **Training mode**: `eps` must be strictly positive (unchanged behavior)
- **Eval mode**: `eps` must be non-negative (`eps=0` is now allowed)
- **Negative `eps`**: always rejected in both modes

cc @mikaylagawarecki (who confirmed this approach sounds reasonable in the issue)

## Test plan

Verified locally with monkeypatching that:
- `eps=0.0, training=False` — works correctly, no error
- `eps=0.0, training=True` — raises `ValueError` with updated message
- `eps=-1.0` — always raises `ValueError` regardless of mode
- `eps=1e-5` — works in both modes (existing behavior preserved)

Fixes #173337

### CODE REVIEW

**Review by mikaylagawarecki — APPROVED** [review]
Thanks
